routing-ortools/main.py

```diff
@@ -32,46 +36,308 @@ def main():
     )
     args = parser.parse_args()
 
-    # Read input data, solve the problem and write the solution.
+    # Read and prepare the input data.
     input_data = read_input(args.input)
+    apply_defaults(input_data)
+    validate_input(input_data)
+    process_distance_matrix(input_data)
+    process_duration_matrix(input_data)
+
+    # Solve the problem and write the output.
     log("Solving routing problem:")
-    log(f"  - num_vehicles: {input_data.get('num_vehicles', 0)}")
-    log(f"  - stops: {len(input_data.get('distance_matrix', []))-1}")
+    log(f"  - vehicles: {len(input_data.get('vehicles', []))}")
+    log(f"  - stops: {len(input_data.get('stops', []))}")
     log(f"  - max duration: {args.duration} seconds")
     solution = solve(input_data, args.duration)
     write_output(args.output, solution)
 
 
+def apply_defaults(input_data: dict[str, Any]) -> None:
+    """
+    Applies default values to the vehicles and stops
+    (if they are given and not already set on them directly).
+    """
+    if "defaults" not in input_data:
+        return input_data
+    defaults = input_data["defaults"]
+    if "vehicles" in defaults:
+        for vehicle in input_data["vehicles"]:
+            for key, value in defaults["vehicles"].items():
+                if key not in vehicle:
+                    vehicle[key] = value
+    if "stops" in defaults:
+        for stop in input_data["stops"]:
+            for key, value in defaults["stops"].items():
+                if key not in stop:
+                    stop[key] = value
+
+
+def _valid_location(element: dict[str, Any]) -> bool:
```

Comment:
What is the difference here between preceding a function with `_` vs not? I believe convention is to use `_` for private methods that shouldn’t be visible outside the module. For this simple example I would say that methods don’t necessarily need to be preceded by `_`.

Reply:
I agree. This file just grew into considerable complexity, so, this was an attempt to slightly distinguish between auxiliary and more core funcs. However, let me dial that back and re-organize the file a bit (addressing above comment as well).